libraries/rush-lib/src/logic/operations/ConsoleTimelinePlugin.ts

```diff
@@ -0,0 +1,244 @@
+ */
+export function _printTimeline(terminal: ITerminal, result: IExecutionResult): void {
+  //
+  // Gather the operation records we'll be displaying. Do some inline max()
+  // finding to reduce the number of times we need to loop through operations.
+  //
+
+  const durationByPhase: Map<IPhase, number> = new Map();
+
+  const data: ITimelineRecord[] = [];
+  let longestNameLength: number = 0;
+  let longestDurationLength: number = 0;
+  let allStart: number = Infinity;
+  let allEnd: number = -Infinity;
+  let workDuration: number = 0;
+
+  for (const [operation, operationResult] of result.operationResults) {
+    if (operation.runner?.silent) {
+      continue;
+    }
+
+    const { stopwatch } = operationResult;
+
+    const { startTime, endTime } = stopwatch;
+
+    if (startTime && endTime) {
+      const nameLength: number = operation.name?.length || 0;
+      if (nameLength > longestNameLength) {
+        longestNameLength = nameLength;
+      }
+
+      const { duration } = stopwatch;
+      const durationString: string = duration.toFixed(1);
+      const durationLength: number = durationString.length;
+      if (durationLength > longestDurationLength) {
+        longestDurationLength = durationLength;
+      }
+
+      if (endTime > allEnd) {
+        allEnd = endTime;
+      }
+      if (startTime < allStart) {
+        allStart = startTime;
+      }
+      workDuration += duration;
+
+      const { associatedPhase } = operation;
+
+      if (associatedPhase) {
+        durationByPhase.set(associatedPhase, (durationByPhase.get(associatedPhase) || 0) + duration);
+      }
+
+      data.push({
+        startTime,
+        endTime,
+        durationString,
+        name: operation.name!,
+        status: operationResult.status
+      });
+    }
+  }
+
+  data.sort((a, b) => a.startTime - b.startTime);
+
+  //
+  // Determine timing for all tasks (wall clock and execution times)
+  //
+
+  const allDuration: number = allEnd - allStart;
+  const allDurationSeconds: number = allDuration / 1000;
+
+  //
+  // Do some calculations to determine what size timeline chart we need.
+  //
+
+  const maxWidth: number = process.stdout.columns || TIMELINE_WIDTH;
+  const chartWidth: number = maxWidth - longestNameLength - longestDurationLength - 4;
+  //
+  // Loop through all operations, assembling some statistics about operations and
+  // phases, if applicable.
+  //
+
+  const busyCpus: number[] = [];
+  function getOpenCPU(time: number): number {
+    const len: number = busyCpus.length;
+    for (let i: number = 0; i < len; i++) {
+      if (busyCpus[i] <= time) {
+        return i;
+      }
+    }
+    return len;
+  }
+
+  // Start with a newline
+  terminal.writeLine('');
+  terminal.writeLine('='.repeat(maxWidth));
+
+  for (const { startTime, endTime, durationString, name, status } of data) {
+    // Track busy CPUs
+    const openCpu: number = getOpenCPU(startTime);
+    busyCpus[openCpu] = Math.max(startTime, endTime - 10);
```

Comment:
Explain the 10-ms fudge factor.

Reply:
Don't actually need this anymore. Turns out there was a bug in Async.forEachAsync (which I fixed both here and in another PR) that was causing the concurrency to exceed the expected value.